File: NightgamesMod/nightgames/match/Participant.java
package nightgames.match;

import nightgames.actions.Action;
import nightgames.actions.Move;
import nightgames.areas.Area;
import nightgames.characters.Attribute;
import nightgames.characters.Character;
import nightgames.characters.State;
import nightgames.global.Global;
import nightgames.items.Item;
import nightgames.status.Stsflag;

import java.util.*;
import java.util.stream.Collectors;

public class Participant {
    private Character character;
    private int score = 0;
    private int roundsToWait = 0;

    // Participants this participant has defeated recently.  They are not valid targets until they resupply.
    private Set<Participant> invalidTargets = new HashSet<>();

    public Participant(Character c) {
        this.character = c;
    }

    Participant(Participant p) {
        try {
            this.character = this.getCharacter().clone();
        } catch (CloneNotSupportedException e) {
            throw new RuntimeException(e);
        }
        this.score = p.score;
        this.invalidTargets = p.invalidTargets;
        this.roundsToWait = p.roundsToWait;
    }

    public Character getCharacter() {
        return character;
    }

    int getScore() {
        return score;
    }

    void incrementScore(int i) {
        score += i;
    }

    void defeated(Participant p) {
        assert !invalidTargets.contains(p);
        invalidTargets.add(p);
        incrementScore(1);
    }

    public Participant copy() {
        return new Participant(this);
    }

    void allowTarget(Participant p) {
        invalidTargets.remove(p);
    }

    public void place(Area loc) {
        character.location.set(loc);
        loc.place(this);
        if (loc.name.isEmpty()) {
            throw new RuntimeException("empty location");
        }
    }

    public boolean canStartCombat(Participant p2) {
        return character.eligible(p2.character);
    }

    public interface ActionCallback {
        Action.Aftermath execute(Action a);
    }


    private Collection<Item> craftItems() {
        int roll = Global.random(15);
        if (character.check(Attribute.Cunning, 25)) {
            if (roll == 9) {
                return List.of(Item.Aphrodisiac, Item.DisSol);
            } else if (roll >= 5) {
                return List.of(Item.Aphrodisiac);
            } else {
                return List.of(Item.Lubricant, Item.Sedative);
            }
        } else if (character.check(Attribute.Cunning, 20)) {
            if (roll == 9) {
                return List.of(Item.Aphrodisiac);
            } else if (roll >= 7) {
                return List.of(Item.DisSol);
            } else if (roll >= 5) {
                return List.of(Item.Lubricant);
            } else if (roll >= 3) {
                return List.of(Item.Sedative);
            } else {
                return List.of(Item.EnergyDrink);
            }
        } else if (character.check(Attribute.Cunning, 15)) {
            if (roll == 9) {
                return List.of(Item.Aphrodisiac);
            } else if (roll >= 8) {
                return List.of(Item.DisSol);
            } else if (roll >= 7) {
                return List.of(Item.Lubricant);
            } else if (roll >= 6) {
                return List.of(Item.EnergyDrink);
            }
        } else {
            if (roll >= 7) {
                return List.of(Item.Lubricant);
            } else if (roll >= 5) {
                return List.of(Item.Sedative);
            }
        }
        return List.of();
    }

    private Collection<Item> searchItems() {
        int roll = Global.random(10);
        switch (roll) {
            case 9:
                return List.of(Item.Tripwire, Item.Tripwire);
            case 8:
                return List.of(Item.ZipTie, Item.ZipTie, Item.ZipTie);
            case 7:
                return List.of(Item.Phone);
            case 6:
                return List.of(Item.Rope);
            case 5:
                return List.of(Item.Spring);
            default:
                return List.of();
        }
    }


    public void move() {
        character.displayStateMessage(character.location.get().getTrap(this));
        var possibleActions = new ArrayList<Action>();
        possibleActions.addAll(character.location.get().possibleActions(this));
        possibleActions.addAll(character.getItemActions());
        possibleActions.addAll(Global.getMatch().getAvailableActions());
        possibleActions.removeIf(a -> !a.usable(this));
        if (character.state == State.combat) {
            if (!character.location.get().fight.battle()) {
                Global.getMatch().resume();
            }
            return;
        } else if (roundsToWait > 0) {
            roundsToWait--;
            return;
        } else if (this.character.is(Stsflag.enthralled)) {
            character.handleEnthrall(act -> act.execute(this));
            return;
        } else if (character.state == State.shower || character.state == State.lostclothes) {
            character.bathe();
            character.state = State.ready;
            return;
        } else if (character.state == State.crafting) {
            character.craft(craftItems());
            character.state = State.ready;
            return;
        } else if (character.state == State.searching) {
            character.search(searchItems());
            character.state = State.ready;
            return;
        } else if (character.state == State.resupplying) {
            character.resupply();
            return;
        } else if (character.state == State.webbed) {
            character.state = State.ready;
            return;
        } else if (character.state == State.masturbating) {
            character.masturbate();
            character.state = State.ready;
            return;
        }
        character.move(possibleActions, character.location.get().encounter(this), act -> act.execute(this));
    }

    public void flee(Area area) {
        var options = character.location.get().possibleActions(this);
        var destinations = options.stream()
                .filter(action -> action instanceof Move)
                .map(action -> (Move) action)
                .map(Move::getDestination)
                .collect(Collectors.toList());
        var destination = destinations.get(Global.random(destinations.size()));
        character.notifyFlight(destination);
        character.travel(destination);
        area.endEncounter();
    }

    public void waitRounds(int i) {
        roundsToWait += i;
    }

    public void endOfMatchRound() {
        character.endOfMatchRound();
    }

    public Area getLocation() { return character.location(); }
}

File: NightgamesMod/nightgames/match/Participant.kt
package nightgames.match

import nightgames.areas.Area
import nightgames.areas.Challenge
import nightgames.characters.Character
import nightgames.characters.Trait
import nightgames.combat.Combat
import nightgames.global.Global
import nightgames.match.Action.Ready
import nightgames.match.Encounter.IntrusionOption
import nightgames.match.actions.Move
import nightgames.match.actions.Resupply
import nightgames.modifier.action.DescribablePredicate
import java.util.*

open class Participant {
    // Below, the participant 'p' is the one who holds the state
    interface State {
        fun allowsNormalActions(): Boolean
        fun move(p: Participant)
        val isDetectable: Boolean
        fun eligibleCombatReplacement(encounter: Encounter, p: Participant, other: Participant): Runnable?
        fun ineligibleCombatReplacement(p: Participant, other: Participant): Runnable?
        fun spotCheckDifficultyModifier(p: Participant): Int
        fun sendAssessmentMessage(p: Participant, observer: Character) {
            if (p.character.mostlyNude()) {
                observer.message("She is completely naked.")
            } else {
                observer.message("She is dressed and ready to fight.")
            }
        }
    }

    var character: Character
        protected set
    var intelligence: Intelligence
        private set
    var dialog: Dialog
        private set
    var score = 0
        private set
    @JvmField
    var state: State = Ready()
    @JvmField
    var invalidAttackers: MutableSet<Participant> = HashSet()
    @JvmField
    var challenges: MutableList<Challenge> = ArrayList()
    private val actionFilter: DescribablePredicate<Action.Instance>

    constructor(c: Character, actionFilter: DescribablePredicate<Action.Instance>) {
        character = c
        intelligence = c.makeIntelligence()
        dialog = c.makeDialog()
        this.actionFilter = actionFilter
    }

    internal constructor(p: Participant) {
        try {
            character = p.character.clone()
        } catch (e: CloneNotSupportedException) {
            throw RuntimeException(e)
        }
        intelligence = p.intelligence
        dialog = p.dialog
        score = p.score
        state = p.state
        invalidAttackers = HashSet(p.invalidAttackers)
        challenges = ArrayList(p.challenges)
        actionFilter = p.actionFilter
    }

    fun copy(): Participant {
        return Participant(this)
    }

    fun incrementScore(amt: Int, reason: String?) {
        score += amt
        location.occupants
                .forEach { participant -> participant.character.message(Match.scoreString(character, amt, reason)) }
    }

    fun place(loc: Area) {
        character.location.set(loc)
        loc.place(this)
        if (loc.name.isEmpty()) {
            throw RuntimeException("empty location")
        }
    }

    open fun canStartCombat(p2: Participant): Boolean {
        return !p2.invalidAttackers.contains(this) && p2.state !is Resupply.State
    }

    fun move() {
        character.displayStateMessage(character.location.get().getTrap(this))
        val possibleActions = ArrayList<Action>()
        possibleActions.addAll(character.getItemActions())
        possibleActions.addAll(Global.getMatch().availableActions)
        possibleActions.removeIf { a: Action -> !a.usable(this) }
        val possibleActionInstances = character.location.get().possibleActions(this).toMutableList()
        possibleActionInstances.addAll(
                possibleActions
                    .mapNotNull { action -> action.newInstance(this, character.location()) }
                    .filter(actionFilter::test).toSet()
        )
        character.status
                .map { s: nightgames.status.Status -> s.makeAllowedActionsPredicate(this) }
                .forEach { filter -> possibleActionInstances.removeIf(filter!!) }
        val callback : (Action.Instance) -> Unit = { it.execute() }
        state.move(this)
        if (state.allowsNormalActions()) {
            if (!character.location.get().encounter(this)) {
                intelligence.move(possibleActionInstances, callback)
            }
        }
    }

    fun flee() {
        val options = character.location.get().possibleActions(this)
        val destinations = options.filterIsInstance<Move.Instance>().map { it.destination }
        val destination = destinations[Global.random(destinations.size)]
        travel(destination, "You dash away and escape into the <b>" + destination.name + ".</b>")
    }

    fun endOfMatchRound() {
        character.getTraits().forEach{ trait: Trait ->
            if (trait.status != null) {
                val newStatus = trait.status.instance(character, null)
                if (!character.has(newStatus)) {
                    character.addNonCombat(Status(newStatus))
                }
            }
        }
        character.endOfMatchRound()
    }

    val location: Area
        get() = character.location()!!

    fun travel(dest: Area, message: String?) {
        state = Ready()
        character.location.get().exit(character)
        character.location.set(dest)
        dest.enter(character)
        if (dest.name.isEmpty()) {
            throw RuntimeException("empty location")
        }
        character.notifyTravel(dest, message)
    }

    open fun timePasses() {}
    fun intrudeInCombat(intrusionOptions: Set<IntrusionOption?>?, noneContinuation: Runnable?) {
        intelligence.intrudeInCombat(
            intrusionOptions,
            character.location.get().possibleActions(this).filterIsInstance<Move.Instance>(),
            { it.execute() },
            noneContinuation
        )
    }

    fun invalidateAttacker(victor: Participant) {
        invalidAttackers.add(victor)
    }

    fun finishMatch() {
        character.finishMatch()
        invalidAttackers.clear()
    }

    open fun pointsForVictory(loser: Participant): Int {
        return loser.pointsGivenToVictor()
    }

    protected open fun pointsGivenToVictor(): Int {
        return 1
    }

    fun accept(c: Challenge) {
        challenges.add(c)
    }

    fun evalChallenges(c: Combat?, victor: Character?) {
        for (chal in challenges) {
            chal.check(c, victor)
        }
    }

    fun makeCombatIntelligence(): nightgames.combat.Intelligence {
        return intelligence.makeCombatIntelligence()
    }
}